### ISSUE TITLE
SSL not enabled by default for owncloud 8.1 container

### ISSUE THREAD

**Issue by asridharan** [docker-library/owncloud#23]
Running the latest owncloud container. All operations are allowed over http, but web-server rejects all https connection. On running bash on an existing container found that SSL module had not been enabled. Had to enabled SSL manually. This should be part of the dockerfile.

**Comment by md5** [comment]
@asridharan SSL is always going to have to be a custom configuration since you have to provide your SSL key and certificate. That being said, there should be instructions in the docs for doing so. I think that documentation belongs [here](https://github.com/docker-library/docs/tree/master/php) for the `php` image. Images that derive from `php`, like `owncloud`, should have a pointer to the SSL documentation in the `php` image.

**Comment by asridharan** [comment]
The problem is that ssl is not enabled in the base image by default. So a user will have to enabled it explicitly. Which effectively means commiting on top of the base image and hence not being able to use updates on the base image (say going from 8.1 -> 8.2). 

For e.g. I had to explicitly run :

``` console
a2enmod ssl
a2ensite default-ssl
```

Commit the image on top of the base and mount the keys. This doesn't make any sense and makes the base image pretty much unusable. 

I think SSL configuration should be enabled in the base. Further the keys directory:
/etc/ssl/certs/
Should mounted explicitly by the user.

**Comment by md5** [comment]
Your suggestion will break the image for users who haven't provided a certificate. This is the error they'll receive:

> AH00526: Syntax error on line 32 of /etc/apache2/sites-enabled/default-ssl.conf:
> SSLCertificateFile: file '/etc/ssl/certs/ssl-cert-snakeoil.pem' does not exist or is empty

**Comment by asridharan** [comment]
I understand, and that's the reason I was saying the user needs to provide
a locally mounted volume, with certificates, that maps to /etc/ssl/certs/

Owncloud manual explicitly states that the website should be run over https
(rightly so), and we should enforce this through the image.

The container should make it easier rather than harder to get SSL going.

-Avinash

**Comment by md5** [comment]
Could you provide a link to that part of the Owncloud manual? I wasn't able to find it.

Not saying you're wrong about HTTPS, I just think that having the link and/or a quotation from the relevant section of the manual would be helpful.

**Comment by md5** [comment]
I found this: https://doc.owncloud.org/server/8.1/admin_manual/configuration_server/harden_server.html#use-https

> **Use HTTPS**
> 
> Using ownCloud without using an encrypted HTTPS connection opens up your server to a man-in-the-middle (MITM) attack, and risks the interception of user data and passwords. It is a best practice, and highly recommended, to always use HTTPS on production servers, and to never allow unencrypted HTTP.
> 
> How to setup HTTPS on your Web server depends on your setup; please consult the documentation for your HTTP server. The following examples are for Apache.

**Comment by asridharan** [comment]
Mike,
Sorry not getting your point. Are you saying we shouldn't enable SSL by
default, though we have a way, or are you still against this argument?

**Comment by md5** [comment]
I'm saying that enabling SSL by default will be a breaking change for existing users. It also requires non-trivial configuration to get it working correctly.

I also don't think that `default-ssl.conf` is an adequate out-of-the-box SSL configuration to just turn on for users without them understanding it. The OwnCloud docs even recommend generating a custom SSL configuration in that same section using the [Mozilla SSL Configuration Generator](https://mozilla.github.io/server-side-tls/ssl-config-generator/), which requires making choices about which browsers should be supported (due to protocol and cipher choices).

**Comment by md5** [comment]
As another data point, the `apache2` Debian package that includes `default-ssl` doesn't enable it by default, likely for the same reasons I stated about users needing to understand their SSL config.

**Comment by md5** [comment]
I'll also add that having the `owncloud` container handle its own SSL is not the only way that users might choose to deploy the application. They may want to have another container like [`jwilder/nginx-proxy`](https://github.com/jwilder/nginx-proxy) or [`ehazlett/interlock`](https://github.com/ehazlett/interlock) handling SSL and only have `owncloud` accessible on the Docker bridge network with no publicly exposed ports. Or they may want to have an Elastic Load Balancer terminate SSL in the context of AWS or perhaps a hardware load balancer like an F5 doing the same.

In short, there is no one-size-fits-all approach to securing the container, so I think that documentation is the best that can be done.

As for your concern about upgrading versions, you're going to have to stop the old container and start a new one anyways, so I don't see how it makes it any more difficult to also rebuilt a derived image that runs `a2enmod ssl` and enables an appropriate SSL configuration.

**Comment by asridharan** [comment]
@md5 I don't agree with your points here. As far as the debian apache package  is concerned, I suspect, the reason for disabling SSL by default is not that it might break existing SSL configuration, but rather that apache is a generic webserver that is also (in quiet a few cases) required to host static, public facing pages that don't necessarily need SSL. SSL in the end is compute heavy and having it on all the time doesn't make sense, if it can be avoided.

Owncloud, on the other hand, is a specific webservice for which SSL is a "must". Atleast for any useful deployment. 

As to your concern about breaking existing user configurations: I see containers as consisting of two parts: code and data. The whole point of having a container is to have a consistent, default, environment for users to execute a code, consistently, irrespective of their deployment. The data allows the users to influence the behavior of the container and hence should be fed in and retrieved by the users from outside the container. 

In this instance, I would think the default environment for owncloud should be with SSL enabled. The data part is the configuration for SSL and the keys that  enable SSL to function properly. This I agree is very user dependent and should be allowed to be fed in using mounted volumes. 

I agree with you that this not a one-fits all solution, for e.g. people might want to use nginx as a proxy. But I believe that these would be the exceptions (e.g. web-scale product deployments of owncloud) rather than the norm. That said rather than keeping SSL disabled by default, we should keep it on by default and allow users to turn off SSL using conf files, (configuring virtual hosts maybe ??) or even environment variables if possible.

**Comment by md5** [comment]
How about something like this: https://github.com/appropriate/docker-owncloud/commit/2532f4f73c49a3f07a7880cfefb47a221e1cb186

I only updated the 8.1 entrypoint for ease of review, but I can revise it to apply to all entrypoints if there's buy-in from the official image maintainers. We'd also need to come up with a decent `owncloud-ssl.conf` (possibly a variation on `default-ssl.conf` or the config generated by the Mozilla SSL Configuration Generator). It probably also needs a variable corresponding to `SSLCertificateChainFile` and possibly one for `SSLCACertificateFile`.

**Comment by asridharan** [comment]
I agree. Sounds like a plan. Thanks a lot. Please tell I can help in any
way.

**Comment by md5** [comment]
I've added an SSL config based on the ownCloud hardening docs and the Mozilla SSL Configuration Generator. See this diff: https://github.com/docker-library/owncloud/compare/643c807d8b88c57b5c1bd340503d7e8c33749cb2...appropriate:enable-ssl

I also enabled `mod_env` and `mod_headers` as recommended in the hardening docs.

It looks like Apache 2.4 no longer uses `SSLCertificateChainFile` and will instead accept a certificate chain in the normal `SSLCertificateFile`. I also didn't add `SSLCACertificateFile` since it's only used to validate client certificates which I think is out of scope.

I ended up using `mod_rewrite` instead of `mod_alias` for the port 80 to 443 redirect to allow for the fact that the container may not know its actual hostname. This was also needed to support a `OWNCLOUD_TLS_PORT` environment variable to handle the situation where container port 443 is exposed as a port other than 443 on the host.

@asridharan If you'd like to spend some time testing this configuration, that would be a great help! :+1:

**Comment by md5** [comment]
BTW, here's how I ran the container:

``` console
$ docker run --rm -p 80:80 -p 443:443 --name owncloud --hostname owncloud \
    -e OWNCLOUD_TLS_CERT=owncloud-cert.pem \
    -e OWNCLOUD_TLS_KEY=owncloud-key.pem \
    -v $(pwd)/owncloud-cert.pem:/etc/ssl/certs/owncloud-cert.pem \
    -v $(pwd)/owncloud-key.pem:/etc/ssl/private/owncloud-key.pem \
    appropriate/owncloud
```

**Comment by asridharan** [comment]
@md5  Will definitely give this a shot.

**Comment by asridharan** [comment]
@md5 have you pushed your image to docker hub? Can't seem to find it on docker hub:

``` console
$ docker search appropriate/owncloud
NAME      DESCRIPTION   STARS     OFFICIAL   AUTOMATED
```

``` console
$ docker pull appropriate/owncloud
Pulling repository appropriate/owncloud
Error: image appropriate/owncloud:latest not found
```

**Comment by md5** [comment]
No. I just built it locally with that name. I'd rather not push it to Docker Hub.

Sorry for the confusion.

**Comment by asridharan** [comment]
Can you push it to locally available public repo? Your docker hub repo would be a good place. Without that not sure how to test this?

**Comment by md5** [comment]
Something like the following should work:

``` console
$ git clone -b enable-ssl https://github.com/appropriate/docker-owncloud.git
$ cd docker-owncloud
$ docker build -t owncloud:enable-ssl 8.1
$ openssl req -new -x509 -newkey rsa:2048 -days 365 \
    -keyout owncloud-key.pem \
    -out owncloud-cert.pem \
    -subj '/CN=owncloud'
$ docker run --rm -p 80:80 -p 443:443 --name owncloud --hostname owncloud \
    -e OWNCLOUD_TLS_CERT=owncloud-cert.pem \
    -e OWNCLOUD_TLS_KEY=owncloud-key.pem \
    -v $(pwd)/owncloud-cert.pem:/etc/ssl/certs/owncloud-cert.pem \
    -v $(pwd)/owncloud-key.pem:/etc/ssl/private/owncloud-key.pem \
    owncloud:enable-ssl
```

I'd rather not create repositories on Docker Hub for things I'm not prepared to support as a standalone concern.

**Comment by asridharan** [comment]
Will give this a shot

**Comment by asridharan** [comment]
Have issues connecting to apache using the built container. Got the owncloud:enable-ssl image going. 

To start with one problem was that the variables OWNCLOUD_TLS_CERT and OWNCLOUD_TLS_CERT need to set to absolute paths of the *.pem files:
OWNCLOUD_TLS_CERT=/etc/ssl/certs/owncloud-cert.pem
OWNCLOUD_TLS_KEY =/etc/ssl/private/owncloud-key.pem

else apache doesn't start since it takes non-absolute paths to be *.pem files starting in /etc/apache2/ssl/...

Once I got the envs correct, can see that apache has started correctly in the container. Here is the outptu of apachectl -M:

``` console
root@owncloud-rc-web-8rx2p:/var/www/html# apachectl -M
AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 10.1.18.20. Set the 'ServerName' directive globally to suppress this message
Loaded Modules:
 core_module (static)
 so_module (static)
 watchdog_module (static)
 http_module (static)
 log_config_module (static)
 logio_module (static)
 version_module (static)
 unixd_module (static)
 access_compat_module (shared)
 alias_module (shared)
 auth_basic_module (shared)
 authn_core_module (shared)
 authn_file_module (shared)
 authz_core_module (shared)
 authz_host_module (shared)
 authz_user_module (shared)
 autoindex_module (shared)
 deflate_module (shared)
 dir_module (shared)
 env_module (shared)
 filter_module (shared)
 headers_module (shared)
 mime_module (shared)
 mpm_prefork_module (shared)
 negotiation_module (shared)
 php5_module (shared)
 rewrite_module (shared)
 setenvif_module (shared)
 socache_shmcb_module (shared)
 ssl_module (shared)
 status_module (shared)
```

However _apachectl status_ is just hanging. I tried connecting to the server using 80 and 443 from the container itself (using curl and lynx) and both connections were hanging. Here is the output of running netstat in the container:

``` console
root@owncloud-rc-web-8rx2p:/var/www/html# netstat
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp6     445      0 owncloud-rc-web-8r:http 10.1.18.1:56477         CLOSE_WAIT 
tcp6     296      0 localhost:https         localhost:37404         CLOSE_WAIT 
tcp6      78      0 localhost:https         localhost:54338         CLOSE_WAIT 
tcp6     211      0 owncloud-rc-web-8:https 10.1.18.1:60176         CLOSE_WAIT 
tcp6     445      0 owncloud-rc-web-8r:http 10.1.18.1:56476         CLOSE_WAIT 
tcp6      79      0 owncloud-rc-web-8:https 10.1.18.1:60528         CLOSE_WAIT 
tcp6      75      0 owncloud-rc-web-8r:http owncloud-rc-web-8:60259 CLOSE_WAIT 
tcp6      78      0 localhost:https         localhost:54328         CLOSE_WAIT 
tcp6     150      0 owncloud-rc-web-8:https 10.1.18.1:60370         CLOSE_WAIT 
tcp6       1      0 owncloud-rc-web-8:https 10.1.18.1:34708         CLOSE_WAIT 
tcp6     227      0 localhost:http          localhost:57324         CLOSE_WAIT 
tcp6       1      0 owncloud-rc-web-8r:http 10.1.18.1:56479         CLOSE_WAIT 
tcp6     211      0 owncloud-rc-web-8:https 10.1.18.1:60237         CLOSE_WAIT 
tcp6     227      0 localhost:http          localhost:56872         CLOSE_WAIT 
tcp6      75      0 owncloud-rc-web-8r:http 10.1.18.1:56702         CLOSE_WAIT 
tcp6     296      0 localhost:https         localhost:37233         CLOSE_WAIT 
tcp6      74      0 owncloud-rc-web-8r:http 10.1.18.1:56587         CLOSE_WAIT 
tcp6       1      0 owncloud-rc-web-8r:http 10.1.18.1:56478         CLOSE_WAIT 
tcp6     211      0 owncloud-rc-web-8:https 10.1.18.1:34707         CLOSE_WAIT 
tcp6      74      0 owncloud-rc-web-8r:http 10.1.18.1:56577         CLOSE_WAIT 
tcp6      84      0 owncloud-rc-web-8r:http 10.1.18.1:56603         CLOSE_WAIT 
tcp6     211      0 owncloud-rc-web-8:https 10.1.18.1:34706         CLOSE_WAIT 
tcp6     227      0 localhost:http          localhost:56889         CLOSE_WAIT 
Active UNIX domain sockets (w/o servers)
Proto RefCnt Flags       Type       State         I-Node   Path
unix  2      [ ]         DGRAM                    32717112 /run/systemd/ask-password/sck.582a94150114a12b
root@owncloud-rc-web-8rx2p:/var/www/html# 
```

One thing that could be causing a problem is that I am running owncloud container as a pod in a kubernetes setup. This works fine with the official owncloud container (ofcourse only with HTTP). But I would assume http and https connections from within the container should work.

Here is an output of the environment variables passed to the container:

``` console
root@owncloud-rc-web-8rx2p:/var/www/html# env
HOSTNAME=owncloud-rc-web-8rx2p
OWNCLOUD_WEB_SVC_PORT_80_TCP_PORT=80
KUBERNETES_PORT=tcp://10.0.0.1:443
KUBERNETES_PORT_443_TCP_PORT=443
PHP_INI_DIR=/usr/local/etc/php
OWNCLOUD_MYSQL_PORT=tcp://10.0.0.48:3306
KUBERNETES_SERVICE_PORT=443
OWNCLOUD_WEB_SVC_PORT_80_TCP=tcp://10.0.0.63:80
KUBERNETES_SERVICE_HOST=10.0.0.1
OWNCLOUD_MYSQL_SERVICE_HOST=10.0.0.48
OWNCLOUD_WEB_SVC_PORT_443_TCP=tcp://10.0.0.63:443
OWNCLOUD_MYSQL_PORT_3306_TCP=tcp://10.0.0.48:3306
OWNCLOUD_WEB_SVC_SERVICE_PORT=80
OWNCLOUD_WEB_SVC_SERVICE_PORT_OWNCLOUD_WEB_HTTP=80
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
GPG_KEYS=0BD78B5F97500D450838F95DFE857D9A90D90EC1 6E4F6AB321FDC07F2C332E3AC2BF0BC433CFC8B3
PWD=/var/www/html
OWNCLOUD_MYSQL_PORT_3306_TCP_ADDR=10.0.0.48
OWNCLOUD_WEB_SVC_PORT_443_TCP_ADDR=10.0.0.63
OWNCLOUD_VERSION=8.1.2
OWNCLOUD_TLS_CERT=/etc/ssl/certs/owncloud-cert.pem
OWNCLOUD_WEB_SVC_PORT_80_TCP_ADDR=10.0.0.63
OWNCLOUD_WEB_SVC_PORT_80_TCP_PROTO=tcp
OWNCLOUD_WEB_SVC_PORT_443_TCP_PORT=443
OWNCLOUD_MYSQL_PORT_3306_TCP_PORT=3306
OWNCLOUD_WEB_SVC_SERVICE_HOST=10.0.0.63
SHLVL=1
HOME=/root
KUBERNETES_PORT_443_TCP_PROTO=tcp
OWNCLOUD_TLS_KEY=/etc/ssl/private/owncloud-key.pem
OWNCLOUD_WEB_SVC_PORT_443_TCP_PROTO=tcp
OWNCLOUD_MYSQL_PORT_3306_TCP_PROTO=tcp
OWNCLOUD_WEB_SVC_SERVICE_PORT_OWNCLOUD_WEB_HTTPS=443
PHP_EXTRA_BUILD_DEPS=apache2-dev
OWNCLOUD_WEB_SVC_PORT=tcp://10.0.0.63:80
KUBERNETES_PORT_443_TCP_ADDR=10.0.0.1
KUBERNETES_PORT_443_TCP=tcp://10.0.0.1:443
PHP_VERSION=5.6.12
PHP_EXTRA_CONFIGURE_ARGS=--with-apxs2
OWNCLOUD_MYSQL_SERVICE_PORT=3306
_=/usr/bin/env
root@owncloud-rc-web-8rx2p:/var/www/html# 
```

Could it a problem with the mod_rewrite ?

**Comment by md5** [comment]
> To start with one problem was that the variables OWNCLOUD_TLS_CERT and OWNCLOUD_TLS_CERT need to set to absolute paths of the *.pem files:
> OWNCLOUD_TLS_CERT=/etc/ssl/certs/owncloud-cert.pem
> OWNCLOUD_TLS_KEY =/etc/ssl/private/owncloud-key.pem

The entrypoint has code that's supposed to account for that. It checks to see if the two variables refer to a file on their own and if not, it checks for them under `/etc/ssl/certs` and `/etc/ssl/private`. What error were you getting?

As for `apachectl status` hanging, I'm not sure why that would be. I'd have to dig in to see what it's trying to do. I believe it tries to hit a `/server-status` URL on the server.

You should try running `netstat -tl` to see listening ports, since `netstat` by itself doesn't show them.

**Comment by md5** [comment]
Could you provide your Kubernetes pod definitions?

**Comment by asridharan** [comment]
netstat does show that apache is listening on 80 and 443:

``` console
root@owncloud-rc-web-8rx2p:/var/www/html# netstat -ntl
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp6       0      0 :::80                   :::*                    LISTEN     
tcp6       0      0 :::443                  :::*                    LISTEN   
```

This is the error I get if I set the environment variables with the file name and not the absolute path:

``` console
root@owncloud-rc-web-afhg1:/var/www/html# apachectl -M
AH00526: Syntax error on line 8 of /etc/apache2/sites-enabled/owncloud-ssl.conf:
SSLCertificateFile: file '/etc/apache2/owncloud-cert.pem' does not exist or is empty
Action '-M' failed.
The Apache error log may have more information.
root@owncloud-rc-web-afhg1:/var/www/html# 
```

If I set the environment variables with the absolute path, apache seems to start fine:

``` console
~/stitch-apps/apps/owncloud $ docker logs 70950257e637
Considering dependency setenvif for ssl:
Module setenvif already enabled
Considering dependency mime for ssl:
Module mime already enabled
Considering dependency socache_shmcb for ssl:
Enabling module socache_shmcb.
Enabling module ssl.
See /usr/share/doc/apache2/README.Debian.gz on how to configure SSL and create self-signed certificates.
To activate the new configuration, you need to run:
  service apache2 restart
Enabling site owncloud-ssl.
To activate the new configuration, you need to run:
  service apache2 reload
[Mon Sep 07 05:58:10.903904 2015] [core:warn] [pid 1] AH00111: Config variable ${APACHE_RUN_DIR} is not defined
AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 10.1.18.23. Set the 'ServerName' directive globally to suppress this message
```

Here is the kubernetes replication controller config (replica is set to 1)

``` console
{
    "kind": "ReplicationController",
    "apiVersion": "v1",
    "metadata": {
        "name": "owncloud-rc-web"
    },
    "spec": {
        "template": {
            "metadata": {
                "labels": {
                    "app": "owncloud-web"
                }
            },
            "spec": {
                "containers": [
                {   
                    "resources": {
                        "limits" : {
                            "cpu": 0.5
                        }
                    },
                    "image":"owncloud:enable-ssl",
                    "name": "owncloud-webservice",
                    "ports": [
                    {"containerPort": 443,
                        "name": "owncloud-https" },
                    {"containerPort": 80,
                        "name": "owncloud-http" }
                    ],  
                        "volumeMounts": [
                        {"name": "owncloud-persistent-storage",
                            "mountPath": "/owncloud/data"},
                        {"name": "owncloud-cert",
                            "mountPath":"/etc/ssl/certs/owncloud-cert.pem"},
                        {"name": "owncloud-key",
                            "mountPath":"/etc/ssl/private/owncloud-key.pem"}
                    ],
                    "env":[
                        {
                                "name":"OWNCLOUD_TLS_CERT",
                                "value":"/etc/ssl/certs/owncloud-cert.pem"
                       },
                        {
                                 "name":"OWNCLOUD_TLS_KEY",
                                 "value":"/etc/ssl/private/owncloud-key.pem"
                        }
                    ]
                }
                ],
                "volumes":[
                    {
                        "name":"owncloud-persistent-storage",
                        "hostPath": {
                            "path":"/owncloud/data"
                        }
                    },
                    {
                        "name":"owncloud-cert",
                        "hostPath": {
                            "path":"/home/asridharan/stitch-apps/apps/owncloud/certs/stitch-owncloud-cert.pem"
                        }
                    },
                    {
                        "name":"owncloud-key",
                        "hostPath": {
                            "path":"/home/asridharan/stitch-apps/apps/owncloud/certs/stitch-owncloud-key.pem"
                        }
                    }
                ]
            }
        }
    }
}
```

**Comment by md5** [comment]
I don't understand why you keep showing examples where you're running `apachectl status` or `apachectl -M` from a shell.

How are you starting the container you're running the shell in? The variable logic is only invoked if the `CMD` for the container is something that starts with `apache2*`. If you're starting it with `bash` or something, that logic doesn't get run.

I suppose that logic could be run for any `CMD`, but I wouldn't want the container to die with `exit 1` in the case of a missing cert if you're just running `bash`.

The gating of that logic with the `if [[ "$1" = apache2* ]]` check is similar to how other official images do their variable handling (e.g. `wordpress`).

**Comment by asridharan** [comment]
@md5 I am not starting apache manually through the shell. The owncloud:enable-ssl container starts normally (using docker run through kubernetes), and apache gets started through the entrypoint scripts. 

The apachectl command I run by executing a bash shell on a **"running container"**. For e.g. for the given docker ps command:

``` console
asridharan@stitch-core ~/stitch-apps/apps/owncloud $ docker ps
CONTAINER ID        IMAGE                                        COMMAND                CREATED             STATUS              PORTS               NAMES
70950257e637        owncloud:enable-ssl                          "/entrypoint.sh apac   8 hours ago         Up 8 hours                              k8s_owncloud-webservice.84214151_owncloud-rc-web-zn1jy_default_26cf2bbf-5525-11e5-847f-5254000f8f71_c2b5342e   
f673ad88aa87        gcr.io/google_containers/pause:0.8.0         "/pause"               8 hours ago         Up 8 hours                              k8s_POD.87991597_owncloud-rc-web-zn1jy_default_26cf2bbf-5525-11e5-847f-5254000f8f71_d0736b6f                   
f1ac04c92124        mysql:5.7                                    "/entrypoint.sh mysq   22 hours ago        Up 22 hours                             k8s_mysql.6ac8eb2d_owncloud-rc-h30yu_default_b6322f5d-54a9-11e5-847f-5254000f8f71_3db7ec98                     
3c2daa80ad87        gcr.io/google_containers/pause:0.8.0         "/pause"               22 hours ago        Up 22 hours                             k8s_POD.807deaeb_owncloud-rc-h30yu_default_b6322f5d-54a9-11e5-847f-5254000f8f71_a2d23f4a                       
f51ee9fa76b7        gcr.io/google_containers/hyperkube:v0.21.2   "/hyperkube proxy --   2 weeks ago         Up 2 weeks                              cocky_franklin                                                                                                 
e963105f75df        gcr.io/google_containers/hyperkube:v0.21.2   "/hyperkube schedule   2 weeks ago         Up 2 weeks                              k8s_scheduler.b725e775_k8s-master-127.0.0.1_default_06b6b39e77c0de608e62b61a71d18205_6b294fe5                  
12525c929158        gcr.io/google_containers/hyperkube:v0.21.2   "/hyperkube apiserve   2 weeks ago         Up 2 weeks                              k8s_apiserver.fe80218_k8s-master-127.0.0.1_default_06b6b39e77c0de608e62b61a71d18205_f4ffcc91                   
78233a0a9e22        gcr.io/google_containers/hyperkube:v0.21.2   "/hyperkube controll   2 weeks ago         Up 2 weeks                              k8s_controller-manager.aad1ee8f_k8s-master-127.0.0.1_default_06b6b39e77c0de608e62b61a71d18205_73529e82         
a24b773c4cc7        gcr.io/google_containers/pause:0.8.0         "/pause"               2 weeks ago         Up 2 weeks                              k8s_POD.e4cc795_k8s-master-127.0.0.1_default_06b6b39e77c0de608e62b61a71d18205_ffdfe456                         
7851a5c91048        gcr.io/google_containers/hyperkube:v0.21.2   "/hyperkube kubelet    2 weeks ago         Up 2 weeks       
```

container **70950257e637** is the owncloud:enable-ssl started by kubernetes. However, the docker logs on this container doesn't give too much information. So to check the status of apache running in this container, you can get into an interfactive shell and run apachectl commands in the shell. This works:

``` console
$ docker exec -it 70950257e637 /bin/bash
root@owncloud-rc-web-zn1jy:/var/www/html# 
```

So to clarify, none of the scripts are being run manually here. 

PS: I will launch a separate node with docker alone (without kubernetes) and run the container separately to see if I can get some more debug info.

**Comment by md5** [comment]
The output of `docker inspect 70950257e637` might be helpful. It seems like the additional initialization logic is not being run, hence the attempt by Apache to fall back to `/etc/apache2/owncloud-cert.pem`.

**Comment by asridharan** [comment]
Here is the output for docker inspect:

``` console
 $ docker inspect 70950257e637
[
{
    "Id": "70950257e63745831da3052f1b1709239991f7bb2708c53d81317fab00ee5aba",
    "Created": "2015-09-07T05:56:26.222049887Z",
    "Path": "/entrypoint.sh",
    "Args": [
        "apache2-foreground"
    ],
    "State": {
        "Running": true,
        "Paused": false,
        "Restarting": false,
        "OOMKilled": false,
        "Dead": false,
        "Pid": 15939,
        "ExitCode": 0,
        "Error": "",
        "StartedAt": "2015-09-07T05:56:33.609007473Z",
        "FinishedAt": "0001-01-01T00:00:00Z"
    },
    "Image": "87fbba1b8a66098d4fd6dd7bbccec3109fbf9af6e5ecb4fb05a3b8c2c438e6c0",
    "NetworkSettings": {
        "Bridge": "",
        "EndpointID": "",
        "Gateway": "",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "HairpinMode": false,
        "IPAddress": "",
        "IPPrefixLen": 0,
        "IPv6Gateway": "",
        "LinkLocalIPv6Address": "",
        "LinkLocalIPv6PrefixLen": 0,
        "MacAddress": "",
        "NetworkID": "",
        "PortMapping": null,
        "Ports": null,
        "SandboxKey": "",
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null
    },
    "ResolvConfPath": "/var/lib/docker/containers/f673ad88aa87c64363a7ecd03e126fc377ffdd46042b068d109b93d355ee1e8c/resolv.conf",
    "HostnamePath": "/var/lib/docker/containers/f673ad88aa87c64363a7ecd03e126fc377ffdd46042b068d109b93d355ee1e8c/hostname",
    "HostsPath": "/var/lib/docker/containers/f673ad88aa87c64363a7ecd03e126fc377ffdd46042b068d109b93d355ee1e8c/hosts",
    "LogPath": "/var/lib/docker/containers/70950257e63745831da3052f1b1709239991f7bb2708c53d81317fab00ee5aba/70950257e63745831da3052f1b1709239991f7bb2708c53d81317fab00ee5aba-json.log",
    "Name": "/k8s_owncloud-webservice.84214151_owncloud-rc-web-zn1jy_default_26cf2bbf-5525-11e5-847f-5254000f8f71_c2b5342e",
    "RestartCount": 0,
    "Driver": "btrfs",
    "ExecDriver": "native-0.2",
    "MountLabel": "",
    "ProcessLabel": "",
    "Volumes": {
        "/dev/termination-log": "/var/lib/kubelet/pods/26cf2bbf-5525-11e5-847f-5254000f8f71/containers/owncloud-webservice/70950257e63745831da3052f1b1709239991f7bb2708c53d81317fab00ee5aba",
        "/etc/ssl/certs/owncloud-cert.pem": "/home/asridharan/stitch-apps/apps/owncloud/certs/stitch-owncloud-cert.pem",
        "/etc/ssl/private/owncloud-key.pem": "/home/asridharan/stitch-apps/apps/owncloud/certs/stitch-owncloud-key.pem",
        "/owncloud/data": "/owncloud/data",
        "/var/www/html": "/var/lib/docker/volumes/602ccb3c4ad9f79db40d38495980f237e6990a72a1ed9bcb5216a0890ed49157/_data"
    },
    "VolumesRW": {
        "/dev/termination-log": true,
        "/etc/ssl/certs/owncloud-cert.pem": true,
        "/etc/ssl/private/owncloud-key.pem": true,
        "/owncloud/data": true,
        "/var/www/html": true
    },
    "AppArmorProfile": "",
    "ExecIDs": [
        "7576b8209e392607235f60b15cc28ec3c63b361abcf82eb8e0681adb8b15cf58"
    ],
    "HostConfig": {
        "Binds": [
            "/owncloud/data:/owncloud/data",
            "/home/asridharan/stitch-apps/apps/owncloud/certs/stitch-owncloud-cert.pem:/etc/ssl/certs/owncloud-cert.pem",
            "/home/asridharan/stitch-apps/apps/owncloud/certs/stitch-owncloud-key.pem:/etc/ssl/private/owncloud-key.pem",
            "/var/lib/kubelet/pods/26cf2bbf-5525-11e5-847f-5254000f8f71/containers/owncloud-webservice/70950257e63745831da3052f1b1709239991f7bb2708c53d81317fab00ee5aba:/dev/termination-log"
        ],
        "ContainerIDFile": "",
        "LxcConf": null,
        "Memory": 0,
        "MemorySwap": -1,
        "CpuShares": 512,
        "CpuPeriod": 0,
        "CpusetCpus": "",
        "CpusetMems": "",
        "CpuQuota": 0,
        "BlkioWeight": 0,
        "OomKillDisable": false,
        "Privileged": false,
        "PortBindings": null,
        "Links": null,
        "PublishAllPorts": false,
        "Dns": null,
        "DnsSearch": null,
        "ExtraHosts": null,
        "VolumesFrom": null,
        "Devices": null,
        "NetworkMode": "container:f673ad88aa87c64363a7ecd03e126fc377ffdd46042b068d109b93d355ee1e8c",
        "IpcMode": "container:f673ad88aa87c64363a7ecd03e126fc377ffdd46042b068d109b93d355ee1e8c",
        "PidMode": "",
        "UTSMode": "",
        "CapAdd": null,
        "CapDrop": null,
        "RestartPolicy": {
            "Name": "",
            "MaximumRetryCount": 0
        },
        "SecurityOpt": null,
        "ReadonlyRootfs": false,
        "Ulimits": null,
        "LogConfig": {
            "Type": "json-file",
            "Config": {}
        },
        "CgroupParent": ""
    },
    "Config": {
        "Hostname": "owncloud-rc-web-zn1jy",
        "Domainname": "",
        "User": "",
        "AttachStdin": false,
        "AttachStdout": false,
        "AttachStderr": false,
        "PortSpecs": null,
        "ExposedPorts": {
            "443/tcp": {},
            "80/tcp": {}
        },
        "Tty": false,
        "OpenStdin": false,
        "StdinOnce": false,
        "Env": [
            "OWNCLOUD_TLS_CERT=/etc/ssl/certs/owncloud-cert.pem",
            "OWNCLOUD_TLS_KEY=/etc/ssl/private/owncloud-key.pem",
            "OWNCLOUD_WEB_SVC_PORT_80_TCP_ADDR=10.0.0.188",
            "OWNCLOUD_WEB_SVC_PORT_80_TCP=tcp://10.0.0.188:80",
            "OWNCLOUD_WEB_SVC_PORT_80_TCP_PORT=80",
            "OWNCLOUD_MYSQL_PORT_3306_TCP_ADDR=10.0.0.48",
            "OWNCLOUD_MYSQL_SERVICE_HOST=10.0.0.48",
            "OWNCLOUD_MYSQL_PORT_3306_TCP_PORT=3306",
            "OWNCLOUD_WEB_SVC_SERVICE_PORT_OWNCLOUD_WEB_HTTPS=443",
            "OWNCLOUD_MYSQL_SERVICE_PORT=3306",
            "OWNCLOUD_MYSQL_PORT=tcp://10.0.0.48:3306",
            "OWNCLOUD_WEB_SVC_PORT_443_TCP_PORT=443",
            "KUBERNETES_PORT_443_TCP_PROTO=tcp",
            "OWNCLOUD_MYSQL_PORT_3306_TCP=tcp://10.0.0.48:3306",
            "OWNCLOUD_WEB_SVC_SERVICE_PORT=80",
            "KUBERNETES_PORT_443_TCP=tcp://10.0.0.1:443",
            "KUBERNETES_PORT_443_TCP_PORT=443",
            "KUBERNETES_PORT_443_TCP_ADDR=10.0.0.1",
            "OWNCLOUD_MYSQL_PORT_3306_TCP_PROTO=tcp",
            "OWNCLOUD_WEB_SVC_SERVICE_HOST=10.0.0.188",
            "KUBERNETES_SERVICE_HOST=10.0.0.1",
            "OWNCLOUD_WEB_SVC_SERVICE_PORT_OWNCLOUD_WEB_HTTP=80",
            "OWNCLOUD_WEB_SVC_PORT_443_TCP=tcp://10.0.0.188:443",
            "KUBERNETES_PORT=tcp://10.0.0.1:443",
            "OWNCLOUD_WEB_SVC_PORT_443_TCP_PROTO=tcp",
            "KUBERNETES_SERVICE_PORT=443",
            "OWNCLOUD_WEB_SVC_PORT_443_TCP_ADDR=10.0.0.188",
            "OWNCLOUD_WEB_SVC_PORT=tcp://10.0.0.188:80",
            "OWNCLOUD_WEB_SVC_PORT_80_TCP_PROTO=tcp",
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
            "PHP_INI_DIR=/usr/local/etc/php",
            "PHP_EXTRA_BUILD_DEPS=apache2-dev",
            "PHP_EXTRA_CONFIGURE_ARGS=--with-apxs2",
            "GPG_KEYS=0BD78B5F97500D450838F95DFE857D9A90D90EC1 6E4F6AB321FDC07F2C332E3AC2BF0BC433CFC8B3",
            "PHP_VERSION=5.6.12",
            "OWNCLOUD_VERSION=8.1.2"
        ],
        "Cmd": [
            "apache2-foreground"
        ],
        "Image": "owncloud:enable-ssl",
        "Volumes": {
            "/var/www/html": {}
        },
        "VolumeDriver": "",
        "WorkingDir": "/var/www/html",
        "Entrypoint": [
            "/entrypoint.sh"
        ],
        "NetworkDisabled": false,
        "MacAddress": "",
        "OnBuild": null,
        "Labels": {
            "io.kubernetes.pod.name": "default/owncloud-rc-web-zn1jy"
        }
    }
}
]
```

Just for completeness putting the kubernetes view of the pods here as well:

``` console
asridharan@stitch-core ~/stitch-apps/apps/owncloud/certs $ kubectl get pods
NAME                    READY     STATUS    RESTARTS   AGE
k8s-master-127.0.0.1    3/3       Running   0          14d
owncloud-rc-h30yu       1/1       Running   0          1d
owncloud-rc-web-zn1jy   1/1       Running   0          11h
asridharan@stitch-core ~/stitch-apps/apps/owncloud/certs $ kubectl describe owncloud-rc-web-zn1jy
error: no resource "owncloud-rc-web-zn1jy" has been defined
asridharan@stitch-core ~/stitch-apps/apps/owncloud/certs $ kubectl describe pods owncloud-rc-web-zn1jy
Name:                           owncloud-rc-web-zn1jy
Namespace:                      default
Image(s):                       owncloud:enable-ssl
Node:                           127.0.0.1/127.0.0.1
Labels:                         app=owncloud-web
Status:                         Running
Reason: 
Message:
IP:                             10.1.18.23
Replication Controllers:        owncloud-rc-web (1/1 replicas created)
Containers:
  owncloud-webservice:
    Image:      owncloud:enable-ssl
    Limits:
      cpu:              500m
    State:              Running
      Started:          Sun, 06 Sep 2015 22:56:33 -0700
    Ready:              True
    Restart Count:      0
    Variables:
      OWNCLOUD_TLS_CERT:        /etc/ssl/certs/owncloud-cert.pem
      OWNCLOUD_TLS_KEY:         /etc/ssl/private/owncloud-key.pem
Conditions:
  Type          Status
  Ready         True 
Volumes:
  owncloud-persistent-storage:
    Type:       HostPath (bare host directory volume)
    Path:       /owncloud/data
  owncloud-cert:
    Type:       HostPath (bare host directory volume)
    Path:       /home/asridharan/stitch-apps/apps/owncloud/certs/stitch-owncloud-cert.pem
  owncloud-key:
    Type:       HostPath (bare host directory volume)
    Path:       /home/asridharan/stitch-apps/apps/owncloud/certs/stitch-owncloud-key.pem
```

And the service that exposes these pods:

``` console
$ kubectl describe service owncloud-web-svc
Name:                   owncloud-web-svc
Namespace:              default
Labels:                 app=owncloud-web-svc
Selector:               app=owncloud-web
Type:                   ClusterIP
IP:                     10.0.0.188
Port:                   owncloud-web-http       80/TCP
Endpoints:              10.1.18.23:80
Port:                   owncloud-web-https      443/TCP
Endpoints:              10.1.18.23:443
Session Affinity:       None
No events.
```

Will try to run this in a separate docker only node later today. I am not that worried about the path declarations for cert and key, more about apache actually hanging after starting. Not sure if the two are related though.

**Comment by md5** [comment]
I don't think the two things are related.

How are you testing with `curl`? I'm thinking that the failure to connect is a networking problem. I don't know much about Kubernetes' networking, but if you're trying to connect to a mapped IP address outside the container instead of `localhost`, I could see that having problems when trying to connect from inside the container.

Looking at the Docker `NetworkSettings`, I can see that Docker's built-in networking is not being used.

**Comment by asridharan** [comment]
@md5  haven't figured out why we still need to give absolute paths in the variables OWNCLOUD_TLS_CERT and OWNCLOUD_TLS_KEY , but figured the problem with hanging SSL connections. 

There was no problem with the kubernetes networking (tried it out with official 8.1 and worked fine over http). I had made the naive mistake of giving an openssl KEY with a passphrase to apache. Since, the container was running in a non-interactive mode, and the logging was set to "WARNING", apache modssl was not throwing the log saying that it was waiting on a passphrase. This was leading to all SSL requests queueing up and connections hanging, and eventually timing out.

Once I gave the container a key without a passphrase, was able to get https access working by running the container in kubernetes :).

**Comment by md5** [comment]
@asridharan Glad you figured out the issue with the passphrase.

As for the path thing, it may be best to just drop the path stuff and require the user to provide an absolute path.

**Comment by asridharan** [comment]
I agree. 

A few other observations. 
a) We need to give users the option for overriding the owncloud-ssl.conf . One way of doing this would be ask users to provide their on conf using a volume, and then using a variable in apache.conf to include owncloud-ssl.conf . This would address your concern of breaking existing SSL configuration.

b) Does it make sense to give the option of over-riding the existing apache.conf, again using an environment varialble which can be fed to the apache command in entrypoint? Ofcourse, the user will have to be made aware of setting the document root correctly to /var/www/html in that case. We will also need to skip the logic of checking all the variables that are specific to owncloud-ssl.conf

**Comment by md5** [comment]
There's no need for environment variables to replace `owncloud-ssl.conf` or `apache.conf`. Simply replacing the files directly in a derived Dockerfile or by mounting them in with `-v` should be sufficient.

**Comment by asridharan** [comment]
That was my first thought as well, the only thing that bothered me was that I am not sure how docker handles overlaying files over existing files of containers. It does it for directories, since it just needs to create a new mount point in the overlayed file system, wasn't sure about files. Will give it a shot and see if it works. If this works then yes we don't need to worry about this.

**Comment by beatkuster** [comment]
@md5 
Any plans to bring your enable-ssl branch into the official image? I guess this would be appreciated by a lot of users.

As getting SSL certs just got a lot easier through letsencrypt.org coming out of closed beta, it also might be worth a thought to include its client into the image and trigger/configure it by environment variables. Default could be false to not break anything for users not aware of SSL.

**Comment by md5** [comment]
@beatkuster I haven't seen any feedback from the image maintainers, so I hadn't planned on doing anything more with that branch. If you or someone else wants to pick up the torch and complete the work on that branch and try to get some support for merging it, I have no objections.

**Comment by pierreozoux** [comment]
@beatkuster I think this is just an nginx config.

You can look at my setup here: https://github.com/indiehosters/ownCloud
(I use it to power: https://framadrive.org/ A+ on qualys)

(If you have troubles, do not hesitate to open issues over there :) )

**Comment by beatkuster** [comment]
@md5 Thanks for the reply. I'm not sure if my docker know-how is already up to this challenge. Working on it though.
@pierreozoux Yes, maybe using owncloud-fpm + nginx is a valid option. Especially when its only the config file you have to supply. In md5's branch I saw that for apache he had to enable some additional modules in the docker-entrypoint.sh, etc.

**Comment by tianon** [comment]
I'm honestly pretty strongly -1 on forcing SSL by default.  For most users of the image, I'd instead recommend using something like NGINX in front, and _that's_ where SSL would terminate, not within the image, so forcing users to setup certs for that fairly-common use case is going to be unnecessary busy work with no firm benefit to the user.

On the other hand, I don't think that kind of thing is well documented (or the changes that would need to be made to this image in a `Dockerfile` which is `FROM owncloud:x.y`) in order to accomplish that easily, so I think that kind of change would meet with more favor from the image maintainers.  Thanks!
